include/albatross/src/cereal/suitesparse.hpp

```diff
@@ -0,0 +1,486 @@
+  // Complex matrices store twice as many elements.
+  // https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/dev/CHOLMOD/Core/cholmod_sparse.c#L209
+  return m.nzmax * (m.xtype == CHOLMOD_COMPLEX ? 2 : 1);
+}
+
+template <class Matrix>
+inline std::size_t get_p_element_size_bytes(const Matrix &m) {
+  return m.itype == CHOLMOD_INT ? sizeof(int) : sizeof(SuiteSparse_long);
+}
+
+template <class Matrix>
+inline std::size_t get_integer_size_bytes(const Matrix &m) {
+  return m.itype == CHOLMOD_LONG ? sizeof(SuiteSparse_long) : sizeof(int);
+}
+
+template <class Archive>
+inline void save(Archive &ar, cholmod_sparse const &m,
+                 std::uint32_t version ALBATROSS_UNUSED) {
+  ar(CEREAL_NVP(m.nrow));
+  ar(CEREAL_NVP(m.ncol));
+  ar(CEREAL_NVP(m.nzmax));
+  ar(CEREAL_NVP(m.stype));
+  ar(CEREAL_NVP(m.itype));
+  ar(CEREAL_NVP(m.xtype));
+  ar(CEREAL_NVP(m.dtype));
+  ar(CEREAL_NVP(m.sorted));
+  ar(CEREAL_NVP(m.packed));
+
+  const std::size_t element_size_bytes = get_element_size_bytes(m);
+  const std::size_t x_elements = get_num_x_elements(m);
+  const std::size_t p_size_bytes = get_p_element_size_bytes(m);
+  const std::size_t integer_size_bytes = get_integer_size_bytes(m);
+
+  detail::encode_array(ar, "m.p", m.p, (m.ncol + 1) * p_size_bytes);
+  detail::encode_array(ar, "m.i", m.i, m.nzmax * integer_size_bytes);
+  if (nullptr != m.nz && !m.packed) {
+    detail::encode_array(ar, "m.nz", m.nz, m.ncol * integer_size_bytes);
+  }
+  if (nullptr != m.x) {
+    detail::encode_array(ar, "m.x", m.x, x_elements * element_size_bytes);
+  }
+  // This should only be non-null for "zomplex" elements
+  ALBATROSS_ASSERT(m.z == nullptr);
+}
+
+template <class Archive>
+inline void load(Archive &ar, cholmod_sparse &m,
+                 std::uint32_t version ALBATROSS_UNUSED) {
+  ar(CEREAL_NVP(m.nrow));
+  ar(CEREAL_NVP(m.ncol));
+  ar(CEREAL_NVP(m.nzmax));
+  ar(CEREAL_NVP(m.stype));
+  ar(CEREAL_NVP(m.itype));
+  ar(CEREAL_NVP(m.xtype));
+  ar(CEREAL_NVP(m.dtype));
+  ar(CEREAL_NVP(m.sorted));
+  ar(CEREAL_NVP(m.packed));
+
+  const std::size_t element_size_bytes = get_element_size_bytes(m);
+  const std::size_t x_elements = get_num_x_elements(m);
+  const std::size_t p_size_bytes = get_p_element_size_bytes(m);
+  const std::size_t integer_size_bytes = get_integer_size_bytes(m);
+
+  detail::suitesparse_cereal_realloc(&m.p, m.ncol + 1, p_size_bytes);
+  detail::suitesparse_cereal_realloc(&m.i, m.nzmax, integer_size_bytes);
+  if (!m.packed) {
+    detail::suitesparse_cereal_realloc(&m.nz, m.ncol, integer_size_bytes);
+  }
+  detail::suitesparse_cereal_realloc(&m.x, m.nzmax, element_size_bytes);
+
+  detail::decode_array(ar, "m.p", m.p, (m.ncol + 1) * p_size_bytes);
+  detail::decode_array(ar, "m.i", m.i, m.nzmax * integer_size_bytes);
+  if (!m.packed) {
+    detail::decode_array(ar, "m.nz", m.nz, m.ncol * integer_size_bytes);
+  }
+  // TODO(@peddie): in what cases may `x` be absent?
```

Comment:
is this a concern? should we add a ticket to follow up?

Reply:
I will check.